src/Employer/Employer.Web/Filters/EmployerAccessCheckFilter.cs

```diff
@@ -0,0 +1,264 @@
+
+                if (!_allowedAccountTypes.Any(x => string.Compare(accountDetails.ApprenticeshipEmployerType, x, true) == 0))
+                    throw new BlockedEmployerException($"Unknown account type {accountDetails.ApprenticeshipEmployerType}");
+
+                _employerAccountTypeCookieWriter.WriteCookie(context.HttpContext.Response, userId, employerAccountId, result);
+            }
+            return result;
+        }
+
+        private async Task<bool> WasHandledByLevyDeclaration(
+            ActionExecutingContext context,
+            ActionExecutionDelegate next,
+            string userId,
+            string employerAccountId,
+            bool requestIsForALevyPage)
+        {
+            bool hasLevyDeclaration = false;
+            if (GetCookieValueForUserAndEmployer(
+                context.HttpContext,
+                userId: userId,
+                employerAccountId: employerAccountId,
+                readValue: _levyCookieWriter.GetCookieFromRequest,
+                result: out string hasLevyDeclarationAsString))
+            {
+                hasLevyDeclaration = bool.Parse(hasLevyDeclarationAsString);
+            }
+            else
+            {
+                hasLevyDeclaration = await HasStoredLevyDeclaration(
+                    employerAccountId: employerAccountId,
+                    userId: userId);
+                _levyCookieWriter.WriteCookie(context.HttpContext.Response, userId, employerAccountId, hasLevyDeclaration);
+            }
+
+            // No Levy declaration = not handled
+            if (!hasLevyDeclaration)
+                return false;
+
+            // Don't allow Levy pages if Levy is already declared
+            if (requestIsForALevyPage)
+            {
+                context.Result = new RedirectToRouteResult(RouteNames.Dashboard_Index_Get, new { employerAccountId });
+                return true;
+            }
+
+            // Allow pages when Levy declaration is present
+            await next();
+            return true;
+        }
+
+        private async Task<bool> WasHandledByEoi(
+            ActionExecutingContext context,
+            ActionExecutionDelegate next,
+            string userId,
+            string employerAccountId)
+        {
+            bool hasEoi = false;
+            if (GetCookieValueForUserAndEmployer(
+                context.HttpContext,
+                userId: userId,
+                employerAccountId: employerAccountId,
+                readValue: _eoiCookieWriter.GetCookieFromRequest,
+                result: out string hasEoiAsString))
+            {
+                hasEoi = bool.Parse(hasEoiAsString);
+            }
+            else
+            {
+                hasEoi = await GetEmployerHasEoi(employerAccountId);
+                _eoiCookieWriter.WriteCookie(context.HttpContext.Response, userId, employerAccountId, hasEoi);
+
+                // Why redirect to the dashboard?
```

Comment:
Do we need this comment?

Reply:
Not really, but it made no sense to me and I wanted it on record :)